### ISSUE TITLE
square of square root is identity

### ISSUE THREAD

**Issue by AshishBora** [tensorflow/tensorflow#5562]
```python
import tensorflow as tf

tensor = tf.constant(0.0)
loss = tf.square(tf.sqrt(tensor))
grad = tf.gradients(loss, tensor)

with tf.Session() as sess:
    print grad[0].eval()
```

This gives `nan`. The gradient of `sqrt(x)` with respect to `x` is `inf` at `0` which is likely causing the problem.

Would be useful to have automatic inference for this such that `sqrt` followed by `square` can be removed at least for the purpose of gradient computation. There are other similar cases (e.g. `exp` of `log`) where such inference would help.

**Comment by AshishBora** [comment]
This toy example may be bad because it tries to compute gradient at 0 and square root is not defined below zero.

What I am actually trying to do is closer to this

```python
import tensorflow as tf

tensor1, tensor2 = tf.constant(1.0), tf.constant(1.0)
tensor3, tensor4 = tf.constant(2.0), tf.constant(2.0)
tensor_list = [tensor1 - tensor2, tensor3 - tensor4]
loss = tf.square(tf.global_norm(tensor_list))

grad = tf.gradients(loss, tensor1)[0]
with tf.Session() as sess:
    print grad.eval()
```

**Comment by sherrym** [comment]
@andydavis1 , could you please take a look? Thanks.

**Comment by drpngx** [comment]
I don't think it's right to permit square(sqrt) for semi-negative values, except negative zero (as per https://en.wikipedia.org/wiki/Signed_zero, your first example). At zero, this is a boundary value, and your code should probably not rely on any particular behavior for the gradient, since it's undefined.

**Comment by hholst80** [comment]
> This gives nan. The gradient of sqrt(x) with respect to x is inf at 0 which is likely causing the problem.

I think your second example is just a more complicated instance of the problem you are describing in your first example, cf. the sqrt here https://github.com/tensorflow/tensorflow/blob/master/tensorflow/g3doc/api_docs/python/functions_and_classes/shard0/tf.global_norm.md

One way to attack this problem is if you do your explicit define the desired behavior in the limit case. I think you would have to resort to defining your operator own gradient for that op because `tf.select` are not to be trusted. At least I have had bad experiences with trying to do that. Now I often I tend to use `tf.clip_by_value` to handle these problems. Ugly, but often does the trick.

Side note: We can understand why we are getting a NaN in the gradient backprop computation by applying the chain rule:

> ### Define
loss(x) = sqr(sqrt(x))
loss(y) = sqr(y)
y(x) = sqrt(x)
d{loss}/d{x} = d{loss}/d{y} * d{y}/d{x} = 2 y (1/2) 1/sqrt(x).
### Plug x=0, y(0)=0 into the expression above,
2 (0) (1/2) 1/sqrt(0) = 0 (-Inf) = NaN.
### we could symbolically, for x>0, simplify this to
sqrt(x) 1/sqrt(x) = 1
### and extend that definition to x=0, why not

**Comment by skye** [comment]
Automatically closing due to lack of recent activity. Since this issue is old at this point, please reopen the issue if it still occurs when tried with the latest version of Tensorflow. Thank you.
